### PR TITLE
Feature/combined infrastructure metrics

### PR DESCRIPTION
**Pull request by kantera800**
Add support for AP_METRICS_QUERY and LINK_METRIC_QUERY messages as part of combined infrastructure metrics flow in order to pass controller certification test 5.7.1
Task #209
1. AP_METRICS_QUERY_MESSAGE (0x800B)
- Controller: add handler to wfaca bmlm send msg to agent 
- Agent: add print if msg received
2. LINK_METRIC_QUERY_MESSAGE = 0x5
- Controller: add handler to wfaca bml, send msg to agent 
- Agent: add print if msg received

### CODE REVIEW

File: controller/src/beerocks/master/wfa_ca.cpp [hunk 1]
```diff
@@ -556,6 +556,15 @@ bool wfa_ca::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx, ieee1905_1::eMessag
...
 
     switch (message_type) {
+
+    case ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE: {
+        cmdu_header = cmdu_tx.create(g_mid, ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE);
+        if (!cmdu_header) {
+            LOG(ERROR) << "Failed building message!";
+            return false;
+        }
+        break;
```
Comment (morantr) [inline controller/src/beerocks/master/wfa_ca.cpp]: Since the this command comes with TLVSs on the test sequence, there is no need to generate this message here. It will automatically be created and sent.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 2]
```diff
@@ -4706,6 +4708,14 @@ bool slave_thread::handle_client_capability_query(Socket *sd, ieee1905_1::CmduMe
     return true;
 }
 
+bool slave_thread::handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE , mid=" << std::dec << int(mid);
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: I think it is better to use std::hex for the mid. I will only have 4 figures that way.

**Review by morantr — CHANGES_REQUESTED** [review]

File: controller/src/beerocks/master/wfa_ca.cpp (hunk 1 above)
Comment (kantera800) [reply controller/src/beerocks/master/wfa_ca.cpp]: Sure but this  is confusing , 
I think that all messages with or without TLVs can be handled automatically and  create_cmdu()  needs to handle only CHANNEL_SELECTION_REQUEST_MESSAGE

Comment (morantr) [reply controller/src/beerocks/master/wfa_ca.cpp]: Here I fixed it for you: #258

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 3]
```diff
@@ -4706,6 +4708,28 @@ bool slave_thread::handle_client_capability_query(Socket *sd, ieee1905_1::CmduMe
     return true;
 }
 
+bool slave_thread::handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    
+    auto tlv_type = cmdu_rx.getNextTlvType();
+    while (tlv_type != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp:4901]: There is no point using while loop here since this message must have one tlvApMetricQuery, so if doing addClass of this TLV and it fails, return false.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 4]
```diff
@@ -4706,6 +4708,28 @@ bool slave_thread::handle_client_capability_query(Socket *sd, ieee1905_1::CmduMe
...
+        if (tlv_type == int(wfa_map::eTlvTypeMap::TLV_AP_METRIC_QUERY)) {
+            auto ap_metrics_query_tlv = cmdu_tx.addClass<wfa_map::tlvApMetricQuery>();
+            if (!ap_metrics_query_tlv) {
+                LOG(ERROR) << "addClass tlvApMetricQuery has failed";
+                return false;
+            }
+            LOG(DEBUG) << "Received AP_METRICS_QUERY_MESSAGE , mid=" << std::dec << int(mid)
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: please print mids as hex.

File: framework/tlvf/tlvf_conf.yaml [hunk 5]
```diff
@@ -10,6 +10,7 @@ include_yaml_path: {
   # "wfa_map/",
   "tlvf/wfa_map/eTlvTypeMap.yaml",
   "tlvf/wfa_map/tlvApCapability.yaml",
+  "tlvf/wfa_map/tlvApMetricQuery.yaml",
```
Comment (morantr) [inline framework/tlvf/tlvf_conf.yaml]: This, change, the change of yaml file, and the auto-generated files must be in a separate commit (previous to this commit)

File: framework/tlvf/yaml/tlvf/wfa_map/tlvApMetric.yaml [hunk 6]
```diff
@@ -9,7 +9,7 @@ tlvApMetric:
     _type: eTlvTypeMap
     _value_const: TLV_AP_METRIC 
   length: uint16_t
-  bssid: sAddress
+  bssid: sMacAddr
```
Comment (morantr) [inline framework/tlvf/yaml/tlvf/wfa_map/tlvApMetric.yaml]: I could understand why it is right to replace sAddress to sMacAddr, but I don't understand how it is relevant to this PR, since I don't see you use it. So this change should be in a separate PR.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 7]
```diff
@@ -4730,6 +4734,26 @@ bool slave_thread::handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx
 
 }
 
+bool slave_thread::handle_link_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+
+    auto tlv_type = cmdu_rx.getNextTlvType();
+    while (tlv_type != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp:4920]: IEEE1905.1 says that there should be only one link metric query TLV. So while loop is not necessary here either. 
Use addClass and return false if it fails.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 8]
```diff
@@ -4730,6 +4734,26 @@ bool slave_thread::handle_ap_metrics_query(Socket *sd, ieee1905_1::CmduMessageRx
...
+        if (tlv_type == int(ieee1905_1::eTlvType::TLV_LINK_METRIC_QUERY)) {
+            auto link_metrics_query_tlv = cmdu_tx.addClass<ieee1905_1::tlvLinkMetricQuery>();
+            if (!link_metrics_query_tlv) {
+                LOG(ERROR) << "addClass tlvLinkMetricQuery has failed";
+                return false;
+            }
+            LOG(DEBUG) << "Received LINK_METRIC_QUERY_MESSAGE , mid=" << std::dec << int(mid)
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: please print the mid as hex.

**Review by morantr — CHANGES_REQUESTED** [review]

**Comment by morantr** [comment]
Both commits messages are not descriptives enough.
As well 16bacb5b2110ec7c209d7ab9d06a38bf8e7802ca headline is too long and wrapped.

I remember that this PR should include saving TLVs content in the database and then send a combined message. Since this PR coding work hasn't finished, it should be draft. Please do so for the next time.

Edit: I see now that the you did the rest of the code on another PR. I really think you should have done it all in a single PR.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 3 above)
Comment (kantera800) [reply agent/src/beerocks/slave/son_slave_thread.cpp:4901]: good point , fixed .

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 4 above)
Comment (kantera800) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: done

File: framework/tlvf/yaml/tlvf/wfa_map/tlvApMetric.yaml (hunk 6 above)
Comment (kantera800) [reply framework/tlvf/yaml/tlvf/wfa_map/tlvApMetric.yaml]: Its a small bug related to the PR.
I'm not sure that opening a new PR for name change is a good idea.
but if you already spent the time to comment on that i'll fix it in separate commit .

File: framework/tlvf/tlvf_conf.yaml (hunk 5 above)
Comment (kantera800) [reply framework/tlvf/tlvf_conf.yaml]: noted for next time if i may.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 7 above)
Comment (kantera800) [reply agent/src/beerocks/slave/son_slave_thread.cpp:4920]: done

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 8 above)
Comment (kantera800) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: done

**Comment by kantera800** [comment]
>I'll fix the commit description after you approve the code fix-ups, don't want to mess it up with auto-squash. regarding the PR i thought that it would be more understandable but in the end i needed both PRs for testing so you are probably right .

File: framework/tlvf/tlvf_conf.yaml (hunk 5 above)
Comment (morantr) [reply framework/tlvf/tlvf_conf.yaml]: No, please fix it in this PR.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 3 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp:4901]: After some digging, I see that I was wrong. Apparently, every CMDU could have VS_TLV inside it, and it can be first.
It means that you must use the while loop as you did, and add a handler for VS_TLV.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 7 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp:4920]: After some digging, I see that I was wrong. Apparently, every CMDU could have VS_TLV inside it, and it can be first.
It means that you must use the while loop as you did, and add a handler for VS_TLV.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 9]
```diff
@@ -136,9 +136,10 @@ bool main_thread::init()
             ieee1905_1::eMessageType::CLIENT_STEERING_REQUEST_MESSAGE,
             ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE,
             ieee1905_1::eMessageType::MULTI_AP_POLICY_CONFIG_REQUEST_MESSAGE,
+            ieee1905_1::eMessageType::AP_METRICS_QUERY_MESSAGE,
+            ieee1905_1::eMessageType::LINK_METRIC_QUERY_MESSAGE,
```
Comment (morantr) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:140]: The LINK_METRIC_QUERY_MESSAGE is not related here. It should be in the next commit AFAIU.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 3 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp:4901]: Since Tomer said it has a pending solution for it, we can just leave it as is.

**Review by morantr — CHANGES_REQUESTED** [review]

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (kantera800) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:140]: told it was too late to do changes....fixed :)

**Review by morantr — CHANGES_REQUESTED** [review]
8484f889cde5e389acb7269fa8f008f2b93ba161 commit message headline needs to be changed.
Agent: Add support for AP_METRICS_QUERY_MESSAGE add TLV_AP_METRIC_QUERY. ->
agent: Add support for AP_METRICS_QUERY_MESSAGE

**Review by morantr — APPROVED** [review]

**Review by arnout — APPROVED** [review]
Description of the PR is wrong :-) but OK to merge.

**Comment by kantera800** [comment]
> Description of the PR is wrong :-) but OK to merge.
fixed the description.
